### ISSUE TITLE
Pathfind cell collision

### ISSUE THREAD

**Issue by Bigbeto123** [Scirra/Construct-3-bugs#3786]
Pathfind object often find themselves going inside their "obstacles". It's clearly perceptible in this demonstration file, but a lot more so in my current project... To the point of them literally passing through the center of the obstacle like it was nothing. Which is completely unacceptable for the quality standard I'm aiming towards within the future release of the game.

## Attach a .c3p
[bug.zip](https://github.com/Scirra/Construct-3-bugs/files/4499217/bug.zip)

## Steps to reproduce

1.Create an enemy family w/ pathfind behavior and a player object with 8 d movement
2.Create a sprite obstacle family. add the family as obstacles for the enemie's
3.create an event block at every x sec making the enemies find a path to the player's pos, and then move along it

Enemies will NOT avoid the obstacle hitbox entirely, disregarding its corners and sometimes even more.

Enemies should COMPLETELY avoid going inside the obstacle. As simple as it is, no more no less; it's a PATHFIND behavior with OBSTACLES - yes, having actual obstacles to it would be nice.

## More details

**Affected browsers/platforms:** 

**First affected release:** 

## System details

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 80.0.3987.163
Browser engine: Chromium
Browser architecture: 64-bit
Context: webapp
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1.25
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.163 Safari/537.36
C3 release: r195 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 133 gb
Storage usage (approx): 442 mb (0.3%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (Intel(R) HD Graphics 630 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_texture_filter_anisotropic
KHR_parallel_shader_compile
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
OVR_multiview2
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by Bigbeto123** [comment]
Here's my current project file if you wanna see it happening to its full extent.
You'll find EVERYTHING related to enemy movement inside the group "Enemies" > "EMovement".
(obstacle setup can be found inside "Spawn" > "SpawnEnemies" Function: EnemySetup).

Needless to say, this is completely unacceptable, and I'm expect a full fix and nothing less. This is a premium product (to a certain extent) and it it's more than reasonable to ask for 2000's-old basic functions to actually work as intended.

**Comment by AshleyScirra** [comment]
This is currently by design. The Pathfinding behavior's obstacles do not block movement. They only cause the path to be calculated around them.

The reason for this is similar to my explanation in another issue: it is virtually impossible to guarantee an object will arrive at its destination if it is blocked by solids. For example if you have a low turning speed, Pathfinding might find a path through a narrow gap. However if the object bumps off solids, its turning circle would not allow it to pass through the gap. So it just gets stuck, never arrives, and people would file that as a bug (and a worse bug too). So, similar to the "just allow overlaps to happen" solution to avoiding other moving objects, the behavior just lets objects overlap solids, which guarantees they can arrive. If this becomes a significant problem you can adjust the settings of the Pathfinding behavior to help avoid it, particularly by using the cell border to keep paths further away from obstacles.

You can also try handling solid collisions yourself, such as stopping objects or making them reverse if they hit a solid, but you'll probably quickly find they get stuck, which is exactly why the behavior doesn't handle this.

Pathfinding and moving along paths is a really, really difficult problem. This is not specific to Construct in any way. It's just a really hard game design problem to solve.
